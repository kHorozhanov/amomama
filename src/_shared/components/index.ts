export * from './dialog/Dialog';
export * from './grid/Grid';
export * from './card/Card';
