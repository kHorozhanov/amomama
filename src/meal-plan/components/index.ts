export * from './header/MealPlanHeader';
export * from './resend/MealPlanResend';
export * from './grid/MealPlanGrid';
