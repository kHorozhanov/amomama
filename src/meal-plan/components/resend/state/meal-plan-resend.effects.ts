import { Observable, EMPTY } from 'rxjs';
import { switchMap, catchError, ignoreElements, tap, finalize } from 'rxjs/operators';
import { FormikActions } from 'formik';
import { MealPlanApi } from '../../../../_shared/services/api';

export const mealPlanResendEffect = (
  event$: Observable<[AMOMAMA.DTO.MealPlanResend, FormikActions<AMOMAMA.DTO.MealPlanResend>]>
): Observable<never> => {
  return event$.pipe(
    switchMap(
      ([form, { setSubmitting, setStatus }]): Observable<unknown> =>
        MealPlanApi.resend(form).pipe(
          tap((): void => setStatus('submmitted')),
          catchError(
            (): Observable<never> => {
              setStatus('error');
              return EMPTY;
            }
          ),
          finalize((): void => setSubmitting(false))
        )
    ),
    ignoreElements()
  );
};
