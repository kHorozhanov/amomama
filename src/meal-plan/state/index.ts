export * from './meal-plan.reducer';
