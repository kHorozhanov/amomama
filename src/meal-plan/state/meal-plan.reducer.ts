export interface MealPlanState {
  showResend: boolean;
}
export function mealPlanReducer(state: MealPlanState, updates: Partial<MealPlanState>): MealPlanState {
  return {
    ...state,
    ...updates
  };
}
