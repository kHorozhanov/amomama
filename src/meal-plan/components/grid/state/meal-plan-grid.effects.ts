import { Observable, of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { GridConfig } from '../../../../_shared/components';
import { MealPlanApi } from '../../../../_shared/services/api';

export type MealPlanRequestInfo = AMOMAMA.RequestInfo<GridConfig<AMOMAMA.DTO.MealPlanItem>>;

export const mealPlanGridEffect = (): Observable<MealPlanRequestInfo> => {
  return MealPlanApi.get().pipe(
    map(
      ({ rows, columns, cell }: AMOMAMA.DTO.MealPlan): MealPlanRequestInfo => {
        const data = cell.reduce((dataMap: Map<string, AMOMAMA.DTO.MealPlanItem>, item: AMOMAMA.DTO.MealPlanItem): Map<
          string,
          AMOMAMA.DTO.MealPlanItem
        > => {
          return dataMap.set(`${item.rowId}_${item.columnId}`, item);
        }, new Map());
        return {
          data: {
            rows,
            columns,
            data
          }
        };
      }
    ),
    catchError(
      (error): Observable<MealPlanRequestInfo> => {
        return of({
          error
        });
      }
    )
  );
};
