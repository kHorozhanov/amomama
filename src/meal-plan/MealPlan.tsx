import React from 'react';
import { MealPlanHeader, MealPlanResend, MealPlanGrid } from './components';
import { mealPlanReducer, MealPlanState } from './state';

export function MealPlan(): JSX.Element {
  const initialState: MealPlanState = {
    showResend: true
  };
  const [state, updateState] = React.useReducer(mealPlanReducer, initialState);
  return (
    <div className='container'>
      <MealPlanHeader />
      {state.showResend && <MealPlanResend onClose={(): void => updateState({ showResend: false })} />}
      <MealPlanGrid />
    </div>
  );
}
