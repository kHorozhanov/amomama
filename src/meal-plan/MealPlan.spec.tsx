import React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import { mealPlanReducer, MealPlanState } from './state';
import { MealPlan } from './MealPlan';
import { MealPlanResend } from './components';

describe('<MealPlan />', (): void => {
  let state: MealPlanState;
  let setState: jest.Mock;

  const shallowComponent = (): ShallowWrapper => {
    jest.spyOn(React, 'useReducer').mockReturnValue([state, setState]);
    return shallow(<MealPlan />);
  };

  beforeEach((): void => {
    state = {
      showResend: false
    };
    setState = jest.fn();
  });

  afterEach((): void => {
    jest.clearAllMocks();
  });

  it('should render correctly', (): void => {
    const component = shallowComponent();
    expect(component.debug()).toMatchSnapshot();
  });

  it('should call useReducer', (): void => {
    shallowComponent();
    expect(React.useReducer).toHaveBeenCalledWith(mealPlanReducer, { showResend: true });
  });

  it('should render MealPlanResend when showResend = true', (): void => {
    state.showResend = true;
    const component = shallowComponent();
    expect(component.debug()).toMatchSnapshot();
  });

  it('should call updateState when MealPlanResend onClose fired', (): void => {
    state.showResend = true;
    const component = shallowComponent();
    const mealPlanResend = component.find(MealPlanResend);
    mealPlanResend.prop('onClose')();
    expect(setState).toHaveBeenCalledWith({ showResend: false });
  });
});
