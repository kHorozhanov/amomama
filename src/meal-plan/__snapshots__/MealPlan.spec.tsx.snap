// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MealPlan /> should render MealPlanResend when showResend = true 1`] = `
"<div className=\\"container\\">
  <MealPlanHeader />
  <MealPlanResend onClose={[Function: onClose]} />
  <MealPlanGrid />
</div>"
`;

exports[`<MealPlan /> should render correctly 1`] = `
"<div className=\\"container\\">
  <MealPlanHeader />
  <MealPlanGrid />
</div>"
`;
